src/expr/src/scalar/mod.rs

```diff
@@ -678,80 +750,80 @@ impl MirScalarExpr {
-                    expr1: Box::new(self.take()),
-                    expr2: Box::new(and_term),
-                    func: BinaryFunc::And,
-                };
+                for and_term in intersection.into_iter() {
+                    *self = MirScalarExpr::CallBinary {
+                        expr1: Box::new(self.take()),
+                        expr2: Box::new(and_term),
+                        func: if undistribute_and {
+                            BinaryFunc::And
+                        } else {
+                            BinaryFunc::Or
+                        },
+                    };
+                }
             }
         }
     }
 
-    /// Collects undistributable terms from AND expressions.
-    fn harvest_ands(&mut self, ands: &mut Vec<MirScalarExpr>) {
-        if let MirScalarExpr::CallBinary {
-            expr1,
-            expr2,
-            func: BinaryFunc::And,
-        } = self
-        {
-            expr1.harvest_ands(ands);
-            expr2.harvest_ands(ands);
-        } else {
-            ands.push(self.clone())
+    /// Collects undistributable terms from X expressions.
+    /// If `and`, X is AND. If not `and`, X is OR.
+    fn harvest_operands(&mut self, operands: &mut Vec<MirScalarExpr>, and: bool) {
+        if let MirScalarExpr::CallBinary { expr1, expr2, func } = self {
+            let operator = if and { BinaryFunc::And } else { BinaryFunc::Or };
+            if *func == operator {
+                expr1.harvest_operands(operands, and);
+                expr2.harvest_operands(operands, and);
+                return;
+            }
         }
+        operands.push(self.clone())
     }
 
     /// Removes undistributed terms from AND expressions.
-    fn suppress_ands(&mut self, ands: &[MirScalarExpr]) {
-        if let MirScalarExpr::CallBinary {
-            expr1,
-            expr2,
-            func: BinaryFunc::And,
-        } = self
-        {
-            // Suppress the ands in children.
-            expr1.suppress_ands(ands);
-            expr2.suppress_ands(ands);
-
-            // If either argument is in our list, replace it by `true`.
-            let tru = MirScalarExpr::literal_ok(Datum::True, ScalarType::Bool);
-            if ands.contains(expr1) {
-                *self = std::mem::replace(expr2, tru);
-            } else if ands.contains(expr2) {
-                *self = std::mem::replace(expr1, tru);
+    /// If `and`, X is AND. If not `and`, X is OR.
+    fn suppress_operands(&mut self, operands: &[MirScalarExpr], and: bool) {
+        if let MirScalarExpr::CallBinary { expr1, expr2, func } = self {
+            let operator = if and { BinaryFunc::And } else { BinaryFunc::Or };
+            if *func == operator {
+                // Suppress the ands in children.
+                expr1.suppress_operands(operands, and);
+                expr2.suppress_operands(operands, and);
+
+                // If either argument is in our list, replace it by `true`.
```

Comment:
What we are actually doing here is replacing the conjunction/distinction expression with one its terms. The literal value is just a temporary value that is destroyed immediately. The comment made me think that for `a AND b, [a], true`, we were rewriting `self` as `true AND b`, but we are actually rewriting it as just `b`.